resources/shaders/dof_bokeh.hlsl

```diff
@@ -0,0 +1,154 @@
+#include "dof_properties.hlsl"
+#include "dof_util.hlsl"
+
+Texture2D source : register(t0);
+Texture2D coc_buffer :register(t1);
+RWTexture2D<float4> output : register(u0);
+SamplerState s0 : register(s0);
+SamplerState s1 : register(s1);
+
+static const float PI = 3.141592654f;
+static const float PI2 = 6.283185308f;
+
+cbuffer CameraProperties : register(b0)
+{
+	float f_number;
+	float bokeh_poly_amount;
+	uint num_blades;
+	int enable_dof;
+};
+
+float GetDownSampledCoC(float2 uv, float2 texel_size)
+{
+	float4 offset = texel_size.xyxy * float2(-0.5f, 0.5f).xxyy;
+
+	float coc0 = coc_buffer.SampleLevel(s1, uv + offset.xy, 0);
+	float coc1 = coc_buffer.SampleLevel(s1, uv + offset.zy, 0);
+	float coc2 = coc_buffer.SampleLevel(s1, uv + offset.xw, 0);
+	float coc3 = coc_buffer.SampleLevel(s1, uv + offset.zw, 0);
+
+	float coc_min = min(min(min(coc0, coc1), coc2), coc3);
+	float coc_max = max(max(max(coc0, coc1), coc2), coc3);
+
+	float coc = coc_max >= -coc_min ? coc_max : coc_min;
+	return coc;
+}
+
+// Maps a value inside the square [0,1]x[0,1] to a value in a disk of radius 1 using concentric squares.
+// This mapPIng preserves area, bi continuity, and minimizes deformation.
+// Based off the algorithm "A Low Distortion Map Between Disk and Square" by Peter Shirley and
+// Kenneth Chiu. Also includes polygon morphing modification from "CryEngine3 Graphics Gems"
+// by Tiago Sousa
+float2 SquareToConcentricDiskMapping(float x, float y, float numSides, float polygonAmount)
+{
+	float phi, r;
+
+	// -- (a,b) is now on [-1,1]�2
+	float a = 2.0f * x - 1.0f;
+	float b = 2.0f * y - 1.0f;
+
+	if (a > -b)                      // region 1 or 2
+	{
+		if (a > b)                   // region 1, also |a| > |b|
+		{
+			r = a;
+			phi = (PI / 4.0f) * (b / a);
+		}
+		else                        // region 2, also |b| > |a|
+		{
+			r = b;
+			phi = (PI / 4.0f) * (2.0f - (a / b));
+		}
+	}
+	else                            // region 3 or 4
+	{
+		if (a < b)                   // region 3, also |a| >= |b|, a != 0
+		{
+			r = -a;
+			phi = (PI / 4.0f) * (4.0f + (b / a));
+		}
+		else                        // region 4, |b| >= |a|, but a==0 and b==0 could occur.
+		{
+			r = -b;
+			if (abs(b) > 0.0f)
+				phi = (PI / 4.0f) * (6.0f - (a / b));
+			else
+				phi = 0;
+		}
+	}
+
+	const float N = numSides;
+	float polyModifier = cos(PI / N) / cos(phi - (PI2 / N) * floor((N * phi + PI) / PI2));
```

Comment:
styleguide

Reply:
Also you need to references every single resource you used from the intrawebs and list it with the license. We need to license where we get our code from.